src/import/load-costume.js

```diff
@@ -30,76 +30,151 @@ const loadVector_ = function (costume, costumeAsset, runtime, rotationCenter, op
-        if (costume.bitmapResolution === 1 && !runtime.v2BitmapAdapter) {
-            log.error('No V2 bitmap adapter present; bitmaps may not render correctly.');
-        } else if (costume.bitmapResolution === 1) {
-            runtime.v2BitmapAdapter.convertResolution1Bitmap(src, (error, dataURI) => {
-                if (error) {
-                    log.error(error);
-                } else if (dataURI) {
-                    // Put back into storage
-                    const storage = runtime.storage;
-                    costume.asset = storage.createAsset(
-                        storage.AssetType.ImageBitmap,
-                        storage.DataFormat.PNG,
-                        runtime.v2BitmapAdapter.convertDataURIToBinary(dataURI),
-                        null,
-                        true // generate md5
-                    );
-                    costume.assetId = costume.asset.assetId;
-                    costume.md5 = `${costume.assetId}.${costume.dataFormat}`;
-                }
-                // Regardless of if conversion succeeds, convert it to bitmap resolution 2,
-                // since all code from here on will assume that.
-                if (rotationCenter) {
-                    rotationCenter[0] = rotationCenter[0] * 2;
-                    rotationCenter[1] = rotationCenter[1] * 2;
-                    costume.rotationCenterX = rotationCenter[0];
-                    costume.rotationCenterY = rotationCenter[1];
-                }
-                costume.bitmapResolution = 2;
-                // Use original src if conversion fails.
-                // The image will appear half-sized.
-                imageElement.src = dataURI ? dataURI : src;
-            });
+
+        baseImageElement.addEventListener('load', onLoad);
+        baseImageElement.addEventListener('error', onError);
+        if (costume.textLayerAsset) {
+            textImageElement = new Image();
+            textImageElement.addEventListener('load', onLoad);
+            textImageElement.addEventListener('error', onError);
+            textImageElement.src = costume.textLayerAsset.encodeDataURI();
         } else {
-            imageElement.src = src;
+            loadedOne = true;
+        }
+        baseImageElement.src = costume.asset.encodeDataURI();
+    }).then(imageElements => {
+        const [baseImageElement, textImageElement] = imageElements;
+
+        let canvas = document.createElement('canvas');
+        const scale = costume.bitmapResolution === 1 ? 2 : 1;
+        canvas.width = baseImageElement.width;
+        canvas.height = baseImageElement.height;
+
+        const ctx = canvas.getContext('2d');
+        ctx.drawImage(baseImageElement, 0, 0);
+        if (textImageElement) {
+            ctx.drawImage(textImageElement, 0, 0);
+        }
+        if (scale !== 1) {
+            if (!runtime.v2BitmapAdapter) {
+                return Promise.reject('No V2 Bitmap adapter present.');
```

Comment:
Does it make sense to move this check/fail earlier so that we don't do the canvas draw image stuff unnecessarily?